test.cc

```diff
@@ -275,3 +277,192 @@ TEST(Sherlock, SubscribeToStreamViaHTTP) {
   // TODO(dkorolev): Unregister the exposed endpoint and free its handler. It's hanging out there now...
   // TODO(dkorolev): Add tests that the endpoint is not unregistered until its last client is done. (?)
 }
+
+struct IntKey {
+  int x;
+  IntKey(int x = 0) : x(x) {}
+  int operator()() const { return x; }
+
+  template <typename A>
+  void serialize(A& ar) {
+    ar(CEREAL_NVP(x));
+  }
+
+  bool operator==(const IntKey& rhs) const { return x == rhs.x; }
+  struct HashFunction {
+    size_t operator()(const IntKey& k) const { return static_cast<size_t>(k.x); }
+  };
+};
+
+struct KeyValueEntry {
+  IntKey key_;
+  double value_;
+  // Uncomment the line below to ensure that it doesn't compile.
+  // constexpr static bool allow_nonthrowing_get = true;
+
+  KeyValueEntry() = default;
+  KeyValueEntry(const int key, const double value) : key_(key), value_(value) {}
+
+  const IntKey& key() const { return key_; }
+  void set_key(const IntKey& key) { key_ = key; }
+
+  template <typename A>
+  void serialize(A& ar) {
+    ar(CEREAL_NVP(key_), CEREAL_NVP(value_));
+  }
+};
+
+struct KeyValueSubscriptionData {
+  atomic_size_t seen_;
+  string results_;
+  KeyValueSubscriptionData() : seen_(0u) {}
+};
+
+struct KeyValueAggregateListener {
+  KeyValueSubscriptionData& data_;
+  size_t max_to_process_ = static_cast<size_t>(-1);
+
+  KeyValueAggregateListener() = delete;
+  explicit KeyValueAggregateListener(KeyValueSubscriptionData& data) : data_(data) {}
+
+  KeyValueAggregateListener& SetMax(size_t cap) {
+    max_to_process_ = cap;
+    return *this;
+  }
+
+  bool Entry(const KeyValueEntry& entry, size_t index, size_t total) {
+    static_cast<void>(index);
+    static_cast<void>(total);
+    if (data_.seen_) {
+      data_.results_ += ",";
+    }
+    data_.results_ += Printf("%d=%.2lf", entry.key(), entry.value_);
+    ++data_.seen_;
+    return data_.seen_ < max_to_process_;
+  }
+
+  void Terminate() {
+    if (data_.seen_) {
+      data_.results_ += ",";
+    }
+    data_.results_ += "DONE";
+  }
+};
+
+TEST(Sherlock, NonPolymorphicKeyValueStorage) {
+  typedef sherlock::kv_storage::API<KeyValueEntry> KVS;
+  KVS api("non_polymorphic_kv_storage");
+
+  // Add the first key-value pair.
+  // Use `UnsafeStream()`, since generally the only way to access the underlying stream is to make API calls.
+  api.UnsafeStream().Emplace(2, 0.5);
+
+  while (!api.CaughtUp()) {
+    // Spin lock, for the purposes of this test.
+    // Ensure that the data has reached the the processor that maintains the in-memory state of the API.
+  }
+
+  // Future expanded syntax.
+  std::future<KeyValueEntry> f1 = api.AsyncGet(KVS::T_KEY(2));
+  KeyValueEntry r1 = f1.get();
+  EXPECT_EQ(0.5, r1.value_);
+
+  // Future short syntax.
+  EXPECT_EQ(0.5, api.AsyncGet(KVS::T_KEY(2)).get().value_);
+
+  // Callback version.
+  struct CallBackTest {
+    explicit CallBackTest(int key, double value, bool expect_fail = false) :
```

Comment:
Why not `expect_succeed` instead of `expect_fail`, and why default parameter except for to confuse the reader? :)